strategy_infura_refresh.go

```diff
@@ -0,0 +1,547 @@
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case err := <-chanErr:
+				if err != nil {
+					log.Printf("\terror while polling ASG details: %v", err)
+				}
+				continue
+			}
+		}
+	}()
+
+	// make sure to stop the following call when done
+	subCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	readyInstances, chanErr2 := detectInstancesReady(subCtx, sess, tg, newInstances)
+
+	instanceReady := make([]instance, 0, count)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case err := <-chanErr2:
+			if err != nil {
+				log.Printf("\terror while polling autoscaling or target group reported target health: %v", err)
+			}
+			continue
+		case inst := <-readyInstances:
+			instanceReady = append(instanceReady, inst)
+			log.Printf("\t(%d/%d) Instance %s is ready, enabling scale-in protection ...", len(instanceReady), count, inst.instanceId)
+
+			err := setInstanceProtection(ctx, sess, asgName, []instance{inst})
+			if err != nil {
+				return nil, err
+			}
+			log.Printf("\tScale-in protection enabled on %s", inst.instanceId)
+
+			if len(instanceReady) >= count {
+				return instanceReady, nil
+			}
+		}
+	}
+}
+
+// detectNewInstances poll the ASG to detect new instances. Instances given in currentInstances will not be reported.
+func detectNewInstances(ctx context.Context, sess *session.Session, asgName string, currentInstances []instance) (chan instance, chan error) {
+	knownSet := make(map[string]instance)
+	for _, inst := range currentInstances {
+		knownSet[inst.instanceId] = inst
+	}
+
+	chanOut := make(chan instance)
+	chanErr := make(chan error)
+
+	period := 20 * time.Second
+
+	go func() {
+		defer close(chanOut)
+		defer close(chanErr)
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.After(period):
+			}
+
+			asg, err := getASGDetails(ctx, sess, asgName)
+			if err != nil {
+				chanErr <- err
+				// don't stop on error, let the caller decide what to do
+				continue
+			}
+
+			for _, inst := range asg.instances {
+				if _, has := knownSet[inst.instanceId]; !has {
+					chanOut <- inst
+					knownSet[inst.instanceId] = inst
+				}
+			}
+		}
+	}()
+
+	return chanOut, chanErr
+}
+
+// detectInstancesReady poll the ASG to detect when instances passed via the newInstances channel are considered healthy.
+// If a target group is also given, it will also be polled for instance healthiness, which will ultimately query the instance custom health-check.
+func detectInstancesReady(ctx context.Context, sess *session.Session, tg *targetGroup, newInstances chan instance) (chan instance, chan error) {
+	chanOut := make(chan instance)
+	chanErr := make(chan error)
+
+	period := 20 * time.Second
+
+	go func() {
+		defer close(chanOut)
+		defer close(chanErr)
+
+		instanceSet := make(map[string]instance)
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case inst := <-newInstances:
```

Comment:
Once `newInstances` is closed by `detectNewInstances` won't this continuously return a zero-value instance object?

For example: https://play.golang.org/p/9ZLMeNYlNkY

Reply:
Yeah good point. I suppose most of the time `<-ctx.Done()` would trigger first and exist but that's a race at least.